MPRIS Integration

jostrander opened the pull request:
Just dropping this in as a start to mpris (#298) for anyone to continue working on or test out. It works using `dbus-send` on my system and currently broadcasts on `org.mpris.MediaPlayer2.google-play-music-desktop-player` with title, artist, album, albumArt, length, etc.


<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/marshallofsound/google-play-music-desktop-player-unofficial-/368)
<!-- Reviewable:end -->

MarshallOfSound commented on src/main/features/linux/mprisService.js:
```diff
@@ -0,0 +1,60 @@
+const mpris = require('mpris-service');
```
Please use the `import` syntax

```js
import mpris from 'mpris-service';
```

It's just the ES6 way :+1:

MarshallOfSound commented on src/main/features/linux/mprisService.js:
```diff
@@ -0,0 +1,60 @@
+        this.listeners();
+    }
+
+    listeners() {
```
More of a style than anything else but "private" methods I.e. methods that are only called internally.  Can they start with an `_` :+1: 

So this would be `_listeners()`

MarshallOfSound commented on src/main/features/linux/mprisService.js:
```diff
@@ -0,0 +1,60 @@
+        this.player.metadata = {
+            // This is required in MPRIS, but no uuid for each track is given by gmusic library.
+            //'mpris:trackid': player.objectPath('track/0'),
+            //'mpris:length' : newSong.length, // 
```
For song length playbackAPI has a `on('time')` I think or something similar

jostrander replied:
What unit of seconds is that using? mpris is expecting micro seconds.

MarshallOfSound commented on src/main/features/linux/mprisService.js:
```diff
@@ -0,0 +1,60 @@
+    updateMetadata(newSong) {
+        this.player.metadata = {
+            // This is required in MPRIS, but no uuid for each track is given by gmusic library.
+            //'mpris:trackid': player.objectPath('track/0'),
```
Hmmmm, can we use the `last.fm` service to pull a UUID or can this just be any old UUID?

Something like `Date.now()`?

jostrander replied:
Honestly I'm not sure it's entirely required, after reviewing docs here:

https://www.freedesktop.org/wiki/Specifications/mpris-spec/metadata/#index1h4

It seems to be relevant when using TrackList only, which I didn't implement because gmusic doesn't implement playlist features.

MarshallOfSound commented:
Woo hoo, cool things are happening.

But the CI is failed :cry:

jostrander commented:
Another issue I noticed with CI, this now requires node-dbus, but node-dbus is a linux specific package. Any ideas on how to only require it for linux?

jostrander commented:
whoops, wrong button

MarshallOfSound commented:
See the `optionalDependencies` section of the `package.json`

Anything in there will only be installed if it can be, but won't throw an error if it can't :+1: 

Add the entire `mpris` package into that section

jostrander commented:
Also ci/circle CI is failing because of eslint, it hates me. Is there a CS tool I can use to clean that file up?

MarshallOfSound commented:
If you use atom there is an eslint plugin that I use.  Highlights things red as you type them to show you that something is wrong.

MarshallOfSound commented:
@jostrander Might be worth giving this a once over

https://github.com/twolfson/google-music-electron/blob/master/lib/mpris.js

jostrander commented:
I took a look at the code there, I implemented the seek and playerStatus, however the desktopEntry is what seems to be needed in order for the Ubuntu sound menu to correctly work. Unfortunately it seems the mpris-service library hasn't implemented something correctly because asking for it's desktopEntry returns "Not Supported". Making manual changes to the mpris-service currently to see if its something I can fix upstream.

jostrander commented:
I do want to point out, even with MPRIS being 90% implemented, media keys still do not work on my system. There seems to be a work around for the spotify client in linux as well that uses XBindKeys and MPRIS:

https://wiki.archlinux.org/index.php/spotify#D-Bus

jostrander commented:
@MarshallOfSound You want to test this out and see if there's anything I've missed. It seems to be working for all features except volume level :+1:. Just going to leave that to the OS to handle. I've removed any mention of media keys because that's not something listed in the spec for mpris unfortunately. :-1:

jostrander commented:
Resolved merge conflicts from gnome media keys.

MarshallOfSound commented on src/main/features/linux/mprisService.js:
```diff
@@ -0,0 +1,84 @@
+  });
+
+  player.on('quit', () => {
+    app.quit();
```
Something to think about here, should this act as a quit, even if we have minimize to task bar turned on?

(I'm not sure how this quit command would be received)

jostrander replied:
This quit should be received only if the mpris client sends a quit, generally speaking the client might want to switch to a different media source, so quitting should in fact quit. The spec does say that this should cause the client to close, but doesn't have to be implemented.

MarshallOfSound replied:
Fair enough :+1:

MarshallOfSound commented:
Sorry @jostrander I forgot where we got to on this.  Is it good to go?

jostrander commented:
Yes this is good to go, just have to fix the merge conflict. The only thing I notice could be an issue is building with node 4.4.0 and needing to use electron-rebuild because this uses dbus as well indirectly.

MarshallOfSound commented:
Ahh never pushed that Linux electron rebuild script thing.  It's sitting in a VM somewhere.  I'll push that when I get home from holiday

jostrander commented:
Well, rebased and fixed the conflict, which was the dbus change from package.json, which might change again when you push your rebuild script, but if that happens tag me and I'll update again.

mcg commented:
Seems like this is broken again(at least media keys in 16.04) since 3.2.5?

MarshallOfSound commented:
I'd it worked in 3.2.4 (it did) it should work in 3.2.5

MarshallOfSound commented:
Nothing changed
